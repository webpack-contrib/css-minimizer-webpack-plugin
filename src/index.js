import * as os from "os";

import { SourceMapConsumer } from "source-map";
import { validate } from "schema-utils";
import serialize from "serialize-javascript";
import pLimit from "p-limit";
import { Worker } from "jest-worker";

import {
  cssnanoMinify,
  cssoMinify,
  cleanCssMinify,
  esbuildMinify,
} from "./utils";

import * as schema from "./options.json";
import { minify as minifyFn } from "./minify";

const warningRegex = /\s.+:+([0-9]+):+([0-9]+)/;

class CssMinimizerPlugin {
  constructor(options = {}) {
    validate(schema, options, {
      name: "Css Minimizer Plugin",
      baseDataPath: "options",
    });

    const {
      minify = cssnanoMinify,
      minimizerOptions,
      test = /\.css(\?.*)?$/i,
      warningsFilter = () => true,
      parallel = true,
      include,
      exclude,
    } = options;

    this.options = {
      test,
      warningsFilter,
      parallel,
      include,
      exclude,
      minify,
      minimizerOptions,
    };
  }

  static isSourceMap(input) {
    // All required options for `new SourceMapConsumer(...options)`
    // https://github.com/mozilla/source-map#new-sourcemapconsumerrawsourcemap
    return Boolean(
      input &&
        input.version &&
        input.sources &&
        Array.isArray(input.sources) &&
        typeof input.mappings === "string"
    );
  }

  static buildWarning(
    warning,
    file,
    sourceMap,
    requestShortener,
    warningsFilter
  ) {
    let warningMessage =
      typeof warning === "string"
        ? warning
        : `${warning.plugin ? `[${warning.plugin}] ` : ""}${
            warning.text || warning.message
          }`;
    let locationMessage = "";
    let source;

    if (sourceMap) {
      const match = warningRegex.exec(warning);

      if (match) {
        const line = +match[1];
        const column = +match[2];
        const original = sourceMap.originalPositionFor({
          line,
          column,
        });

        if (
          original &&
          original.source &&
          original.source !== file &&
          requestShortener
        ) {
          ({ source } = original);

          warningMessage = `${warningMessage.replace(warningRegex, "")}`;
          locationMessage = `${requestShortener.shorten(original.source)}:${
            original.line
          }:${original.column}`;
        }
      }
    }

    if (warningsFilter && !warningsFilter(warning, file, source)) {
      return null;
    }

    const builtWarning = new Error(
      `${file} from Css Minimizer plugin\n${warningMessage}${
        locationMessage ? ` ${locationMessage}` : ""
      }`
    );

    builtWarning.name = "Warning";
    builtWarning.hideStack = true;
    builtWarning.file = file;

    return builtWarning;
  }

  static buildError(error, file, sourceMap, requestShortener) {
    let builtError;

    if (typeof error === "string") {
      builtError = new Error(`${file} from Css Minimizer plugin\n${error}`);
      builtError.file = file;

      return builtError;
    }

    if (error.line) {
      const original =
        sourceMap &&
        sourceMap.originalPositionFor({
          line: error.line,
          column: error.column,
        });

      if (original && original.source && requestShortener) {
        builtError = new Error(
          `${file} from Css Minimizer plugin\n${
            error.message
          } [${requestShortener.shorten(original.source)}:${original.line},${
            original.column
          }][${file}:${error.line},${error.column}]${
            error.stack
              ? `\n${error.stack.split("\n").slice(1).join("\n")}`
              : ""
          }`
        );
        builtError.file = file;

        return builtError;
      }

      builtError = new Error(
        `${file} from Css Minimizer plugin\n${error.message} [${file}:${
          error.line
        },${error.column}]${
          error.stack ? `\n${error.stack.split("\n").slice(1).join("\n")}` : ""
        }`
      );
      builtError.file = file;

      return builtError;
    }

    if (error.stack) {
      builtError = new Error(
        `${file} from Css Minimizer plugin\n${error.stack}`
      );
      builtError.file = file;

      return builtError;
    }

    builtError = new Error(
      `${file} from Css Minimizer plugin\n${error.message}`
    );
    builtError.file = file;

    return builtError;
  }

  static getAvailableNumberOfCores(parallel) {
    // In some cases cpus() returns undefined
    // https://github.com/nodejs/node/issues/19022
    const cpus = os.cpus() || { length: 1 };

    return parallel === true
      ? cpus.length - 1
      : Math.min(Number(parallel) || 0, cpus.length - 1);
  }

  async optimize(compiler, compilation, assets, optimizeOptions) {
    const cache = compilation.getCache("CssMinimizerWebpackPlugin");
    let numberOfAssetsForMinify = 0;
    const assetsForMinify = await Promise.all(
      Object.keys(typeof assets === "undefined" ? compilation.assets : assets)
        .filter((name) => {
          const { info } = compilation.getAsset(name);

          if (
            // Skip double minimize assets from child compilation
            info.minimized
          ) {
            return false;
          }

          if (
            !compiler.webpack.ModuleFilenameHelpers.matchObject.bind(
              // eslint-disable-next-line no-undefined
              undefined,
              this.options
            )(name)
          ) {
            return false;
          }

          return true;
        })
        .map(async (name) => {
          const { info, source } = compilation.getAsset(name);

          const eTag = cache.getLazyHashedEtag(source);
          const cacheItem = cache.getItemCache(name, eTag);
          const output = await cacheItem.getPromise();

          if (!output) {
            numberOfAssetsForMinify += 1;
          }

          return { name, info, inputSource: source, output, cacheItem };
        })
    );

    let getWorker;
    let initializedWorker;
    let numberOfWorkers;

    if (optimizeOptions.availableNumberOfCores > 0) {
      // Do not create unnecessary workers when the number of files is less than the available cores, it saves memory
      numberOfWorkers = Math.min(
        numberOfAssetsForMinify,
        optimizeOptions.availableNumberOfCores
      );

      getWorker = () => {
        if (initializedWorker) {
          return initializedWorker;
        }

        initializedWorker = new Worker(require.resolve("./minify"), {
          numWorkers: numberOfWorkers,
          enableWorkerThreads: true,
        });

        // https://github.com/facebook/jest/issues/8872#issuecomment-524822081
        const workerStdout = initializedWorker.getStdout();

        if (workerStdout) {
          workerStdout.on("data", (chunk) => process.stdout.write(chunk));
        }

        const workerStderr = initializedWorker.getStderr();

        if (workerStderr) {
          workerStderr.on("data", (chunk) => process.stderr.write(chunk));
        }

        return initializedWorker;
      };
    }

    const limit = pLimit(
      getWorker && numberOfAssetsForMinify > 0 ? numberOfWorkers : Infinity
    );
    const { SourceMapSource, RawSource } = compiler.webpack.sources;
    const scheduledTasks = [];

    for (const asset of assetsForMinify) {
      scheduledTasks.push(
        limit(async () => {
          const { name, inputSource, cacheItem } = asset;
          let { output } = asset;

          if (!output) {
            let input;
            let inputSourceMap;

            const { source: sourceFromInputSource, map } =
              inputSource.sourceAndMap();

            input = sourceFromInputSource;

            if (map) {
              if (CssMinimizerPlugin.isSourceMap(map)) {
                inputSourceMap = map;
              } else {
                compilation.warnings.push(
                  new Error(`${name} contains invalid source map`)
                );
              }
            }

            if (Buffer.isBuffer(input)) {
              input = input.toString();
            }

            const options = {
              name,
              input,
              inputSourceMap,
              minify: this.options.minify,
              minifyOptions: this.options.minimizerOptions,
            };

            let result;

            try {
              result = await (getWorker
                ? getWorker().transform(serialize(options))
                : minifyFn(options));
            } catch (error) {
              const hasSourceMap =
                inputSourceMap &&
                CssMinimizerPlugin.isSourceMap(inputSourceMap);

              compilation.errors.push(
                CssMinimizerPlugin.buildError(
                  error,
                  name,
                  hasSourceMap
                    ? new SourceMapConsumer(inputSourceMap)
                    : // eslint-disable-next-line no-undefined
                      undefined,
                  // eslint-disable-next-line no-undefined
                  hasSourceMap ? compilation.requestShortener : undefined
                )
              );

              return;
            }

            output = { warnings: [], errors: [] };

            for (const item of result.outputs) {
              if (item.map) {
                let originalSource;
                let innerSourceMap;

                if (output.source) {
                  ({ source: originalSource, map: innerSourceMap } =
                    output.source.sourceAndMap());
                } else {
                  originalSource = input;
                  innerSourceMap = inputSourceMap;
                }

                // TODO need API for merging source maps in `webpack-source`
                output.source = new SourceMapSource(
                  item.code,
                  name,
                  item.map,
                  originalSource,
                  innerSourceMap,
                  true
                );
              } else {
                output.source = new RawSource(item.code);
              }
            }

            if (result.errors && result.errors.length > 0) {
              const hasSourceMap =
                inputSourceMap &&
                CssMinimizerPlugin.isSourceMap(inputSourceMap);

              for (const error of result.errors) {
                output.warnings.push(
                  CssMinimizerPlugin.buildError(
                    error,
                    name,
                    hasSourceMap
                      ? new SourceMapConsumer(inputSourceMap)
                      : // eslint-disable-next-line no-undefined
                        undefined,
                    // eslint-disable-next-line no-undefined
                    hasSourceMap ? compilation.requestShortener : undefined
                  )
                );
              }
            }

            if (result.warnings && result.warnings.length > 0) {
              const hasSourceMap =
                inputSourceMap &&
                CssMinimizerPlugin.isSourceMap(inputSourceMap);

              for (const warning of result.warnings) {
                const buildWarning = CssMinimizerPlugin.buildWarning(
                  warning,
                  name,
                  hasSourceMap
                    ? new SourceMapConsumer(inputSourceMap)
                    : // eslint-disable-next-line no-undefined
                      undefined,
                  // eslint-disable-next-line no-undefined
                  hasSourceMap ? compilation.requestShortener : undefined,
                  this.options.warningsFilter
                );

                if (buildWarning) {
                  output.warnings.push(buildWarning);
                }
              }
            }

            await cacheItem.storePromise({
              source: output.source,
              warnings: output.warnings,
              errors: output.errors,
            });
          }

          if (output.warnings && output.warnings.length > 0) {
            for (const warning of output.warnings) {
              compilation.warnings.push(warning);
            }
          }

          if (output.errors && output.errors.length > 0) {
            for (const error of output.errors) {
              compilation.errors.push(error);
            }
          }

          const newInfo = { minimized: true };
          const { source } = output;

          compilation.updateAsset(name, source, newInfo);
        })
      );
    }

    const result = await Promise.all(scheduledTasks);

    if (initializedWorker) {
      await initializedWorker.end();
    }

    return result;
  }

  apply(compiler) {
    const pluginName = this.constructor.name;
    const availableNumberOfCores = CssMinimizerPlugin.getAvailableNumberOfCores(
      this.options.parallel
    );

    compiler.hooks.compilation.tap(pluginName, (compilation) => {
      compilation.hooks.processAssets.tapPromise(
        {
          name: pluginName,
          stage:
            compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
          additionalAssets: true,
        },
        (assets) =>
          this.optimize(compiler, compilation, assets, {
            availableNumberOfCores,
          })
      );

      compilation.hooks.statsPrinter.tap(pluginName, (stats) => {
        stats.hooks.print
          .for("asset.info.minimized")
          .tap(
            "css-minimizer-webpack-plugin",
            (minimized, { green, formatFlag }) =>
              // eslint-disable-next-line no-undefined
              minimized ? green(formatFlag("minimized")) : ""
          );
      });
    });
  }
}

CssMinimizerPlugin.cssnanoMinify = cssnanoMinify;
CssMinimizerPlugin.cssoMinify = cssoMinify;
CssMinimizerPlugin.cleanCssMinify = cleanCssMinify;
CssMinimizerPlugin.esbuildMinify = esbuildMinify;

export default CssMinimizerPlugin;
