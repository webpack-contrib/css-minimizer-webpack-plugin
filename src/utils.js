/* istanbul ignore next */
async function cssnanoMinify(
  input,
  inputSourceMap,
  minimizerOptions = { preset: "default" }
) {
  const load = async (module) => {
    let exports;

    try {
      // eslint-disable-next-line import/no-dynamic-require, global-require
      exports = require(module);

      return exports;
    } catch (requireError) {
      let importESM;

      try {
        // eslint-disable-next-line no-new-func
        importESM = new Function("id", "return import(id);");
      } catch (e) {
        importESM = null;
      }

      if (requireError.code === "ERR_REQUIRE_ESM" && importESM) {
        exports = await importESM(module);

        return exports.default;
      }

      throw requireError;
    }
  };

  const [[name, code]] = Object.entries(input);
  const postcssOptions = {
    to: name,
    from: name,
    ...minimizerOptions.processorOptions,
  };

  if (typeof postcssOptions.parser === "string") {
    try {
      postcssOptions.parser = await load(postcssOptions.parser);
    } catch (error) {
      throw new Error(
        `Loading PostCSS "${postcssOptions.parser}" parser failed: ${error.message}\n\n(@${name})`
      );
    }
  }

  if (typeof postcssOptions.stringifier === "string") {
    try {
      postcssOptions.stringifier = await load(postcssOptions.stringifier);
    } catch (error) {
      throw new Error(
        `Loading PostCSS "${postcssOptions.stringifier}" stringifier failed: ${error.message}\n\n(@${name})`
      );
    }
  }

  if (typeof postcssOptions.syntax === "string") {
    try {
      postcssOptions.syntax = await load(postcssOptions.syntax);
    } catch (error) {
      throw new Error(
        `Loading PostCSS "${postcssOptions.syntax}" syntax failed: ${error.message}\n\n(@${name})`
      );
    }
  }

  if (inputSourceMap) {
    postcssOptions.map = { annotation: false };
  }

  // eslint-disable-next-line global-require
  const postcss = require("postcss");
  // eslint-disable-next-line global-require
  const cssnano = require("cssnano");
  const result = await postcss([cssnano(minimizerOptions)]).process(
    code,
    postcssOptions
  );

  return {
    code: result.css,
    map: result.map && result.map.toString(),
    warnings: result.warnings().map(String),
  };
}

/* istanbul ignore next */
async function cssoMinify(input, inputSourceMap, minimizerOptions) {
  // eslint-disable-next-line global-require,import/no-extraneous-dependencies
  const csso = require("csso");
  const [[filename, code]] = Object.entries(input);
  const result = csso.minify(code, {
    filename,
    sourceMap: Boolean(inputSourceMap),
    ...minimizerOptions,
  });

  return {
    code: result.css,
    map: result.map && result.map.toJSON(),
  };
}

/* istanbul ignore next */
async function cleanCssMinify(input, inputSourceMap, minimizerOptions) {
  // eslint-disable-next-line global-require,import/no-extraneous-dependencies
  const CleanCSS = require("clean-css");
  const [[name, code]] = Object.entries(input);
  const result = await new CleanCSS({
    sourceMap: Boolean(inputSourceMap),
    ...minimizerOptions,
  }).minify({ [name]: { styles: code } });

  return {
    code: result.styles,
    map: result.sourceMap && result.sourceMap.toJSON(),
    warnings: result.warnings,
  };
}

/* istanbul ignore next */
async function esbuildMinify(input, sourceMap, minimizerOptions) {
  const buildEsbuildOptions = (esbuildOptions = {}) => {
    // Need deep copy objects to avoid https://github.com/terser/terser/issues/366
    return {
      loader: "css",
      minify: true,
      legalComments: "inline",
      ...esbuildOptions,
      sourcemap: false,
    };
  };

  // eslint-disable-next-line import/no-extraneous-dependencies, global-require
  const esbuild = require("esbuild");

  // Copy `esbuild` options
  const esbuildOptions = buildEsbuildOptions(minimizerOptions);

  // Let `esbuild` generate a SourceMap
  if (sourceMap) {
    esbuildOptions.sourcemap = true;
    esbuildOptions.sourcesContent = false;
  }

  const [[filename, code]] = Object.entries(input);

  esbuildOptions.sourcefile = filename;

  const result = await esbuild.transform(code, esbuildOptions);

  return {
    code: result.code,
    // eslint-disable-next-line no-undefined
    map: result.map ? result.map : undefined,
    warnings:
      result.warnings.length > 0
        ? result.warnings.map((item) => {
            return {
              source: item.location && item.location.file,
              line: item.location && item.location.line,
              column: item.location && item.location.column,
              plugin: item.pluginName,
              message: `${item.text}${
                item.detail ? `\nDetails:\n${item.detail}` : ""
              }${
                item.notes.length > 0
                  ? `\n\nNotes:\n${item.notes
                      .map(
                        (note) =>
                          `${
                            note.location
                              ? `[${note.location.file}:${note.location.line}:${note.location.column}] `
                              : ""
                          }${note.text}${
                            note.location
                              ? `\nSuggestion: ${note.location.suggestion}`
                              : ""
                          }${
                            note.location
                              ? `\nLine text:\n${note.location.lineText}\n`
                              : ""
                          }`
                      )
                      .join("\n")}`
                  : ""
              }`,
            };
          })
        : [],
  };
}

export { cssnanoMinify, cssoMinify, cleanCssMinify, esbuildMinify };
