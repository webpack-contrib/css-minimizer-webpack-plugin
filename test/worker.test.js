import serialize from "serialize-javascript";

import { transform } from "../src/minify";

import CssMinimizerPlugin from "../src";

import { normalizeErrors } from "./helpers";

describe("worker", () => {
  it("should minify css", async () => {
    const options = {
      name: "entry.css",
      input: ".foo{color:red;}\n.bar{color:coral;}",
      inputSourceMap: {
        version: 3,
        sources: ["foo.css", "bar.css"],
        names: [],
        mappings: "AAAA,KAAK,iBAAiB,KAAK,UAAU,OAAO",
        file: "x",
        sourcesContent: [".foo{color:red;}", ".bar{color:coral;}"],
      },
      minimizer: {
        implementation: CssMinimizerPlugin.cssnanoMinify,
        options: { discardComments: false },
      },
    };
    const result = await transform(serialize(options));

    expect(result).toMatchSnapshot("result");
  });

  it("should work inputSourceMap as prev", async () => {
    const options = {
      name: "entry.css",
      input: ".foo{color:red;}\n.bar{color:coral;}",
      inputSourceMap: {
        version: 3,
        sources: ["foo.css", "bar.css"],
        names: [],
        mappings: "AAAA,KAAK,iBAAiB,KAAK,UAAU,OAAO",
        file: "x",
        sourcesContent: [".foo{color:red;}", ".bar{color:coral;}"],
      },
      minimizer: {
        implementation: CssMinimizerPlugin.cssnanoMinify,
        options: { discardComments: false },
      },
    };
    const result = await transform(serialize(options));

    expect(result).toMatchSnapshot("result");
  });

  it("should work options.minify function", async () => {
    const options = {
      name: "entry.css",
      input: ".foo{color:red;}\n.bar{color:coral;}",
      minimizer: {
        implementation: () => {
          return { code: ".minify {};" };
        },
        options: { discardComments: false },
      },
    };
    const result = await transform(serialize(options));

    expect(result).toMatchSnapshot("result");
  });

  it("should emit error", async () => {
    const options = {
      name: "entry.css",
      input: false,
      minimizer: {
        implementation: CssMinimizerPlugin.cssnanoMinify,
        options: { preset: "default" },
      },
    };

    try {
      await transform(serialize(options));
    } catch (error) {
      const normalizeError = { ...error };

      normalizeError.message = [error.message.split("\n")];

      expect(normalizeErrors(normalizeError.message)).toMatchSnapshot("error");
    }
  });
});
