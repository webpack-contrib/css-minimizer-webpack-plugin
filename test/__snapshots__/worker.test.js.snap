// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worker should emit error: error 1`] = `
Array [
  "/entry.css:1:1: Unknown word",
]
`;

exports[`worker should minify css: result 1`] = `
Object {
  "errors": Array [],
  "outputs": Array [
    Object {
      "code": ".foo{color:red}.bar{color:coral}",
      "map": Object {
        "file": "entry.css",
        "mappings": "AAAA,KAAK,SAAU,CACf,KAAK,WAAY",
        "names": Array [],
        "sources": Array [
          "entry.css",
        ],
        "sourcesContent": Array [
          ".foo{color:red;}
.bar{color:coral;}",
        ],
        "version": 3,
      },
    },
  ],
  "warnings": Array [],
}
`;

exports[`worker should work inputSourceMap as prev: result 1`] = `
Object {
  "errors": Array [],
  "outputs": Array [
    Object {
      "code": ".foo{color:red}.bar{color:coral}",
      "map": Object {
        "file": "entry.css",
        "mappings": "AAAA,KAAK,SAAU,CACf,KAAK,WAAY",
        "names": Array [],
        "sources": Array [
          "entry.css",
        ],
        "sourcesContent": Array [
          ".foo{color:red;}
.bar{color:coral;}",
        ],
        "version": 3,
      },
    },
  ],
  "warnings": Array [],
}
`;

exports[`worker should work options.minify function: result 1`] = `
Object {
  "errors": Array [],
  "outputs": Array [
    Object {
      "code": ".minify {};",
      "map": undefined,
    },
  ],
  "warnings": Array [],
}
`;
